### DISCUSSION TITLE
How can an Executor communicate task failure to Prefect?

### DISCUSSION THREAD

**Discussion by jbweston** [PrefectHQ/prefect#3390] [General] [answered]
Hi! First off thanks for making Prefect, it's an awesome project.

I have a question about how prefect handles a certain failure mode.

My understanding from reading the `FlowRunner.run` implementation is that (assuming there's no `mapped` edges or anything) it essentially  does a bunch of calls to `Executor.submit`, submitting the tasks in topological order without waiting for any of them to finish, and relying on the Executor to manage task dependencies (as indicated [in the README](https://github.com/PrefectHQ/prefect/blob/master/src/prefect/engine/executors/README.md)).

`FlowRunner.run` then does a [single call](https://github.com/PrefectHQ/prefect/blob/master/src/prefect/engine/flow_runner.py#L600) to `Executor.wait` on all the futures that it needs to determine the final flow state.

**My question is**: how can the _`Executor`_ communicate to Prefect that a task was unable to run for some reason?

For example, the Dask `Executor` [makes a call to `client.gather`](https://github.com/PrefectHQ/prefect/blob/master/src/prefect/engine/executors/dask.py#L451), and this method will AFAIK raise an exception if one of the `gather`ed Dask tasks fail too many times. This could happen if the usercode (`Task.run`) fails in a way that Prefect does not handle (e.g. Segfault, SIGKILL, OOM, ...).

Another example would be a hypothetical executor that runs jobs on some cloud computing platform "directly". If you run out of money / allotted computing time for the month then how can the executor communicate an `OutOfMoney` error to Prefect?

It seems to me that API specified in the [Executor README](https://github.com/PrefectHQ/prefect/blob/master/src/prefect/engine/executors/README.md) simply does not allow for this possibility.

**Comment by cicdw** [comment, marked as answer]
Hi @jbweston - this is such a great question!  You are correct that there isn't really a first-class way to capture this failure mode using Prefect Core alone.  This is precisely where Prefect Server / Cloud shine - in that case, if any tasks are mid run and killed, they will stop sending a heartbeat to the API.  [The Zombie Killer](https://docs.prefect.io/orchestration/concepts/services.html#zombie-killer) will eventually mark such a task as "Failed", from which point either the full flow run will resume (via [Lazarus](https://docs.prefect.io/orchestration/concepts/services.html#lazarus)) or retry (if the task has retry settings configured).

**Reply by jbweston** [reply]
@cicdw thanks for the rapid response!

I'll be sure to check out the capabilities of Prefect Server.
